tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -958,7 +975,565 @@ int main(int argc, char **argv)
+     *= type=test
+     *# Servers MUST ensure that they negotiate the
+     *# same cipher suite when receiving a conformant updated ClientHello (if
+     *# the server selects the cipher suite as the first step in the
+     *# negotiation, then this will happen automatically).
+     **/
+     {
+         DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+         EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN,
+                                                        S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+
+         DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+         EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+         EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+         EXPECT_SUCCESS(s2n_config_disable_x509_verification(config));
+
+         DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+         EXPECT_NOT_NULL(server_conn);
+         EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+         EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+         DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+         EXPECT_NOT_NULL(server_conn);
+         EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+         EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+         struct s2n_test_io_pair io_pair = { 0 };
+         EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+         EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+         /* Force the HRR path */
+         client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+         /* Send ClientHello */
+         s2n_blocked_status blocked = 0;
+         EXPECT_OK(s2n_negotiate_until_message(client_conn, &blocked, SERVER_HELLO));
+
+         /* Receive ClientHello */
+         EXPECT_OK(s2n_negotiate_until_message(server_conn, &blocked, HELLO_RETRY_MSG));
+
+         /* Rearrange the cipher preference order so that a different cipher will be
+          * picked by the server */
+         server_conn->config->security_policy->cipher_preferences->suites[0] = &s2n_tls13_aes_256_gcm_sha384;
+
+         EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                S2N_ERR_BAD_MESSAGE);
+     }
+
+     /**
+     *= https://tools.ietf.org/rfc/rfc8446#4.1.4
+     *= type=test
+     *# The value of selected_version in the HelloRetryRequest
+     *# "supported_versions" extension MUST be retained in the ServerHello,
+     *# and a client MUST abort the handshake with an "illegal_parameter"
+     *# alert if the value changes.
+     **/
+     {
```

Comment:
How is this test different from the one on line 1234? I'm getting a little lost.

Reply:
The 1234 test ([now deleted](https://github.com/aws/s2n-tls/pull/3337/commits/d5d63e1319b8f0d22a1b33c23955e0f665c4aefd)) was changing the legacy version sent by the server after receiving the first client hello, and then checking for a failure, as this would indicate the supported versions extension function had been called.

In this test, it is ensured that the client aborts the handshake if the selected version in the first server hello (the hello retry request) differs from the selected version in the second server hello, sent after receiving the second client hello.

So, this test negotiates until the server is going to send the second server hello, sets the selected version to something not in the hello retry server hello, and then expects a failure when the negotiation continues and the client receives the second server hello.